r1cs/core/src/constraint_system.rs

```diff
@@ -431,6 +431,55 @@ impl<F: Field> ConstraintSystem<F> {
         self.mode == SynthesisMode::Debug
     }
 
+    /// Return the number of non-zero entries of matrix `A`.
+    pub fn num_non_zero_matrix_a(&self) -> usize {
+        Self::num_non_zero_helper(&self.at)
+    }
+
+    /// Return the number of non-zero entries of matrix `B`.
+    pub fn num_non_zero_matrix_b(&self) -> usize {
+        Self::num_non_zero_helper(&self.bt)
+    }
+
+    /// Return the number of non-zero entries of matrix `C`.
+    pub fn num_non_zero_matrix_c(&self) -> usize {
+        Self::num_non_zero_helper(&self.ct)
+    }
+
+    /// Return the maximum number of non-zero entries among matrices `A`, `B`, and `C`.
+    pub fn num_non_zero(&self) -> usize {
+        let max = *[
+            self.num_non_zero_matrix_a(),
+            self.num_non_zero_matrix_b(),
+            self.num_non_zero_matrix_c(),
+        ]
+        .iter()
+        .max()
+        .expect("iterator is not empty");
+        max
+    }
+
+    /// A simple function that balances the non-zero entries between A and B.
+    pub fn balance_matrices(&mut self) {
```

Comment:
Does it make sense to call this function always (not only from Marlin index function) ? It's a question that makes sense also considering a future merging of the 2 matrix representations

Reply:
Or, let me put it in another way, can we completely kill the actual matrix representation as we have in ginger-lib (we don't have groth anymore) and just keep the Marlin one ?

Reply:
This function should be called after the construction of the R1CS matrices is completed. At the moment a R1CSs would not know when to call `balance_matrices()` (unless it calls it each time a constraint is added, but it would be inefficient). If we want to achieve what you propose in the comment, I see two options:
- Adding a `finalize()` function to `ConstraintSystem` interface, which could then call the balancer. In this case I would also opt to make `at`, `bt` and `ct` private fields, and provide getter functions for them, which would return an error if called before the call to `finalize()`
- Removing `balance_matrices()` altogether and instead balance the constraints as we insert them to the R1CS. However this strategy would lead to a worse matrix balancing (even though I expect that in typical use cases, it would still perform quite well).